README.md


# React native formly documentation 
|  | Status |
| - | - |
| Dependencies | [![Dependencies](https://img.shields.io/david/irohitb/rn-formly)](https://david-dm.org/irohitb/rn-formly) [![Dev dependencies](https://img.shields.io/david/dev/irohitb/rn-formly)](https://david-dm.org/irohitb/rn-formly) [![Peer dependencies](https://img.shields.io/david/peer/irohitb/rn-formly)](https://david-dm.org/irohitb/rn-formly)|
| Package | [![npm package version](https://img.shields.io/npm/v/rn-formly)](https://www.npmjs.com/package/rn-formly) [![npm downloads](https://img.shields.io/npm/dt/rn-formly)](https://www.npmjs.com/package/rn-formly)

### About

rn-formly is a light-weight React Native implementation of a formly component. 

By using rn-formly you can easily create beautiful forms, signup screens, settings (upcoming in v2) in less than 2 minutes by just passing a JSON. 

RN-formly  supports plugging in custom validators, loading data asynchronously, passing custom errors, progress bar, loading spinner,  handles safe area for both iOS and Android and comes with almost all the popular component you would require to build amazing forms

For the Advance user, rn-formly allows you to customize the styling of individual component as well (see component props below for more information)


### Note

If you find this repo interesting do not forgot to give it a star. 

### Screenshots 
 <img src="https://user-images.githubusercontent.com/32276134/72256135-7e162f00-362e-11ea-9500-45b552e27660.png" width=250 height=400> <img src="https://user-images.githubusercontent.com/32276134/72256136-7e162f00-362e-11ea-96ff-6da347fae333.png" width=250 height=400>
<img src="https://user-images.githubusercontent.com/32276134/72256137-7e162f00-362e-11ea-971f-bab0288d8955.png" width=250 height=400>
<img src="https://user-images.githubusercontent.com/32276134/72256138-7e162f00-362e-11ea-91e7-1cb2c325c736.png" width=250 height=400> <img src="https://user-images.githubusercontent.com/32276134/72256139-7eaec580-362e-11ea-89cc-80ed6fa3fe5e.png" width=250 height=400>
<img src="https://user-images.githubusercontent.com/32276134/72256141-7eaec580-362e-11ea-9805-7945e5c52d70.png" width=250 height=400>


<img src="https://user-images.githubusercontent.com/32276134/72256142-7eaec580-362e-11ea-9738-efe7ce49e8a4.png" width=250 height=400> <img src="https://user-images.githubusercontent.com/32276134/72256143-7f475c00-362e-11ea-9074-932075147bcc.png" width=250 height=400>
<img src="https://user-images.githubusercontent.com/32276134/72256143-7f475c00-362e-11ea-9074-932075147bcc.png" width=250 height=400>

## Installation 


RN-Formly use following as peerDependency 
1. react-native-device-info
2. react-native-image-crop-picker


```
npm i react-native-image-crop-picker react-native-device-info rn-formly--save
```


For iOS, If you are using **React Native => 0.60** then simple do 

```
pod install 
```


else follow this [guide](https://github.com/react-native-community/react-native-device-info)


### Setting up image crop picker

#### IOS setup


Image Crop picker requires you to have relevant permission to access camera and storage for both iOS and Android 

1. Go to info.plist and add this 

```
<plist version="1.0">
  <dict>
    ...
    <key>NSPhotoLibraryUsageDescription</key>
    <string>$(PRODUCT_NAME) would like access to your photo gallery</string>
    <key>NSCameraUsageDescription</key>
    <string>$(PRODUCT_NAME) would like to use your camera</string>
    <key>NSPhotoLibraryAddUsageDescription</key>
    <string>$(PRODUCT_NAME) would like to save photos to your photo gallery</string>
    <key>NSMicrophoneUsageDescription</key>
    <string>$(PRODUCT_NAME) would like to use your microphone (for videos)</string>
  </dict>
</plist>
```

You refer to following [documentation](https://github.com/react-native-community/react-native-image-picker/blob/master/docs/Install.md) for more information (and probably for integration in complex project) 

or look into this [commit](https://github.com/irohitb/formly-example/commit/2814782d01f6f111208089ff4628c28752919d8d)



#### Android Setup 

Add following lines in your android/app/src/main/AndroidManifest.xml

```
 <uses-permission android:name="android.permission.INTERNET" />
 <uses-permission android:name="android.permission.CAMERA" />
 <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>
```

look at this [commit](https://github.com/irohitb/formly-example/commit/939b81a90c088ed141b12af9ee8fc7340068f106)

If this does not work, refer to following instructions. 

1. [For Image Croper](https://github.com/react-native-community/react-native-image-picker/blob/master/docs/Install.md)
2. [For React Native Device Info](https://github.com/react-native-community/react-native-device-info)


## Usage 

To get started/or a try it out, you can clone or check out the following [repo](https://github.com/irohitb/formly-example)

**Minimilistic Setup Example** 

```
import SignupFormComponent from 'rn-formly';


const inputFields = [
  {
    key: 'name', 
    type: 'text', 
    label: `Your Full Name`,
    required: true,
    helper: 'Using your real name would make it more likely for you to get a match',
    templateOptions: { 
      componentProps: {
        placeholder: 'Frank Murphy'
      }
    }
  }, 
  {
    key: 'otp', 
    type: 'otp', 
    label: 'Enter OTP',
    helper: '(optional) but try to enter incorrect date', 
    templateOptions: {
      noOfTextInput: 5,
    }
  }]
  
  const App: () => React$Node = () => {

  const justLogDataFromForms = (index, key, currentValue,  payload) => {
     console.log('Logging from Parent log')
    console.log(index, key, currentValue, payload)
  }

  return (
    <>
      <SignupFormComponent 
      inputFields={inputFields}
      globalButtonText={'Next'}
      onButtonClick={justLogDataFromForms}
      defaultColor={'green'}
      progressBarProps={{
        blink: false
      }}
      /> 
    </>
  );
};

```

### Required Props 

| **SNO** | **Prop** | **Type** | **Default** | **Required** | **description** | 
|----------|----------|-------------|--------------|-------------|----------|
| 1. | inputFields | Array | Nope | Yes | [Click here](https://github.com/irohitb/rn-formly#inputfields-props) |
| 2. | onButtonClick | Async function | Nope | Yes | Passed function receives `index` of the element in array, `key` associated with it and the complete payload, This is triggered after user have clicked on the Next button, and before the iteration happens, you can throw errors here to prevent from incrementing to the text state |

This is how error handling is done behind the scene: ` if (error.message) setErrorData({status: true, message: error.message})`

### Optional Props

Important Props are highlighted.

#### General Optional Props 

| **SNO** | **Prop** | **Type** | **Default** | **Required** | **description** | 
|----------|----------|-------------|--------------|-------------|----------|
| 1. | textStyle | stylesheet object | none | No | Heading text style (style for your `label` key in `inputFields` props) | 
| 2. | progressBar | object | blink: false | No | click here to know more about ProgressBar Component | 
| 3. | helperTextStyle | stylesheet object | none | No | Helper text style (style for your `helper` key in `inputFields` )
| 4. |  backgroundViewColor | string | 'white' | No | background for the view of your form | 
| **5.** | **defaultColor** | **string**  | **'black'** | **no** | **Formats button color, text color and sub component color accordingly (can be override by passing styles for individual component)** | 
| **6.** | **onFinish** | **function** | **none** | **No** | **Function which is triggered after user have itterated over all the elements in the passed inputFields** |
| 7. | backIconStyle | stylesheet object | none | No | Styling for the back button `<` | 

#### Button Optional props 

Validation and Error throwing are done in `inputFields` array. See [inputFields](https://github.com/irohitb/rn-formly#inputfields-props) section to know more

| **SNO** | **Prop** | **Type** | **Default** | **Required** | **description** | 
|----------|----------|-------------|--------------|-------------|----------|
| 1. | buttonSelectedStyle | stylesheet object | none | No | Styling of the button when a user is allowed to move to the next field |
| 2. | buttonNotSelectedStyle | stylesheet object | none | No | styling of the button when a user is not allowed to move to the next field |
| 3. | buttonSelectedTextStyle | stylesheet object | none | No | styling of the text inside the button when a user is allowed to move to the next field | 
| 4. | buttonNotSelectedTextStyle | stylesheet object | none | no | styling of the text inside the button when a user is not allowed to move to the next field |

#### Error Props 

Validation and Error throwing is done in the `inputFields` array. See `inputFeilds` section to know more.

| **SNO** | **Prop** | **Type** | **Default** | **Required** | **description** | 
|----------|----------|-------------|--------------|-------------|----------|
| 1. | defaultErrorMessage | string | 'Sorry Something went wrong' | No | Error message for user when error occurs but the error does not contain a message for failure |
| 2. | errorStyle | stylesheet object | none | No | Text Styling for error message | 


#### Note 

1. for `stylesheet object` type props. you need to pass props like this 

```
const styles = StyleSheet.create({
  someStyle: {
    ....
    ....
  }, 
})
```

and then pass `styles.someStyle` in props 

2. For `stylesheet object,` Default `None` does not mean that element won't have any styles; the props provided by the user would override the default styling.

## inputFields Props 


inputFields Props supports most of the popular components you would need in a form. If we are missing any component, create an issue, and we should be able to add within 2-4 days. 


### Currently Supported Component 

1. [Text](https://github.com/irohitb/rn-formly#text-component) 
2. [dataTyper](https://github.com/irohitb/rn-formly#datetyper)
3. [checkboxes](https://github.com/irohitb/rn-formly#checkboxes)
4. [picker]((https://github.com/irohitb/rn-formly#picker))
5. [image](https://github.com/irohitb/rn-formly#image-single-image),
6. [images](https://github.com/irohitb/rn-formly#images)
7. [AutoComplete](AutoComplete)
8. [otp](https://github.com/irohitb/rn-formly#otp) 

### Structuring your Array Object 

```
  {
    key: 'name', 
    type: 'text', 
    label: 'Your Full Name',
    helper: 'Using your real name would make it more likely for you to get a match',
    templateOptions: { 
      componentProps: {
        placeholder: 'Frank Murphy'
      },
      templateStyle: styles.textStyle // refer to the style component
    }
  }
 ```
 
 #### Global Properties for each object
 
| S.NO | key | type | description | Required  |
| ---- | --- | ------ | ----------- | --------  |
| 1. | key   | string | Unique key to identify the object  | Yes |
| 2. | type  | string | type of component you want to use (should be one of the above) | Yes | 
| 3. | label | string | Heading for the field | Yes | 
| 4. | helper | string | Helper for the field | No | 
| 5. | templateOptions | object | props for specific components | No | 
| 6. | Validator | function | sometimes you want to do validation in the real-time and not after the next button is clicked, in this case, pass the validator function. A validator function would receive the following data in its argument `index, key, currentData, payload` | No |
| 7. | required | Boolean | if `required: true` the field cannot be empty i.e. button to iterate to the next component will be disabled | no | 


### Text Component 

```
{
    key: 'name', 
    type: 'text', 
    label: 'Your Full Name',
    helper: 'Using your real name would make it more likely for you to get a match',
    templateOptions: { 
      componentProps: {
        placeholder: 'Frank Murphy'
      },
      templateStyle: styles.textStyle // refer to the style component
    }
  }
  ```
Component level Props are mostly passed in `templateOptions`

##### TemplateOptions keys with value

| S.NO | key | type | description | Required  |
| ---- | --- | ---- | ----------- | --------  |
| 1. | componentProps | object | Can pass all the props for `TextInput ` | no | 
| 2. | templateStyle | style object | stying for text input | no | 





### dateTyper

```
  {
    key: 'dob', 
    type: 'dateTyper', //change this to Dob component
    label: 'Your Date of birth',
    helper: 'Your Birthdate will help us in connecting you with people of similar age',
    required: true
  }, 
  ```
  
##### TemplateOptions key/value

| S.NO | key | type | description | Required  | Default | 
| ---- | --- | ---- | ----------- | --------  | ------- |
| 1. | dateFormat | string | should be either `DDMMYYYY`, or `MMDDYYYY` or `YYYYMMDD` | No |


### checkboxes

```
  {
    key: "gender", 
    type: 'checkboxes',
    label: 'Select your Gender', 
    helper: '(Optional), Please select a gender', 
    templateOptions: {
      options: [
  {
  key: 'male',
  label: 'Male',
  value: false 
  },
  {
    key: 'female', 
    label: 'Female',
    value: false
  },
  {
    key: 'others', 
    label: 'Others',
    value: false
  }
]
    }
  }, 
  ```
##### TemplateOptions key/value

| S.NO | key | type | description | Required  | Default | 
| ---- | --- | ---- | ----------- | --------  | ------- |
| 1. | options | Array | should contain `key`, `label` and `value` | Yes | None


### image (single image)

```
  {
    key: 'image',
    type:'image', 
    label: "Upload your cover image", 
    helper: 'You can change your profile pic anytime from settings',
    templateOptions: {
     cropHeight: 200,
     cropWidth: 300
    }
  },

```

##### TemplateOptions key/value

| S.NO | key | type | description | Required  | Default | 
| ---- | --- | ---- | ----------- | --------  | ------- |
| 1. | cropWidth | number | width of the image user is allowed to crop/select | no | none |
| 2. | cropHeight | number | heoght of the image user is allowed to crop/select | no | none |


### images

```
  {
    key: 'image',
    type:'images', 
    label: "Upload your cover image", 
    helper: 'You can change your profile pic anytime from settings',
    templateOptions: {
     cropHeight: 200,
     cropWidth: 300
    }
  },

```

##### TemplateOptions key/value

| S.NO | key | type | description | Required  | Default | 
| ---- | --- | ---- | ----------- | --------  | ------- |
| 1. | cropWidth | number | width of the images user is allowed to crop/select | no | none |
| 2. | cropHeight | number | heoght of the images user is allowed to crop/select | no | none |


### OTP

```
   {
    key: 'otp', 
    type: 'otp', 
    label: 'Enter OTP',
    helper: '(optional) but try to enter incorrect date', 
    templateOptions: {
      noOfTextInput: 5,
    }
  }, 

```

##### TemplateOptions key/value

| S.NO | key | type | description | Required  | Default | 
| ---- | --- | ---- | ----------- | --------  | ------- |
| 1. | noOfTextInput | number | number of character in your otp  | yes | none |

### AutoComplete

```
{
    key: 'autoComplete', 
    type: 'autoComplete', 
    label: 'Type Programming language',
    helper: "This is an example of Auto Complete Component, Don't try to make sense out of the returned data",
    templateOptions: { 
      asyncFunction: async function (text) {
        return async data
      }
    }
  }
```

##### TemplateOptions key/value
| S.NO | key | type | description | Required  | Default | 
| ---- | --- | ---- | ----------- | --------  | ------- |
| 1. | componentProps |  object | Can pass all the props for `TextInput ` | no | 
| 2. | asyncFunction | async function | this function will recive the input text and should return suggestions accordingly (via making some api call etc)  | yes | none | 
| 3. | loaderRequired | Boolean | shows loader until user recieves the data from async function | no | false | 
| 4. | listViewStyle | styling object | styling for the list list view | no | nope |
| 5. | listTextStyle | styling object | styling for the text in the list view | no | nope |


### Picker 

```
  {
    key: "gender", 
    type: 'picker',
    label: 'Select your Gender', 
    helper: '(Optional), Please select a gender', 
    templateOptions: {
      options: [
  {
  key: 'male',
  label: 'Male',
  value: false 
  },
  {
    key: 'female', 
    label: 'Female',
    value: false
  },
  {
    key: 'others', 
    label: 'Others',
    value: false
  }
]
    }
  }, 
  ```
##### TemplateOptions key/value

| S.NO | key | type | description | Required  | Default | 
| ---- | --- | ---- | ----------- | --------  | ------- |
| 1. | options | Array | should contain `key`, `label` and `value` | Yes | None


CHANGELOG.md
# Version 1.0.15 (2020-01-29)
1. Fixed form Progress bug 
2. Updated docs

index.js
import React, {useState, useEffect} from 'react' 
import {View, Text, StyleSheet, TouchableOpacity, TextInput, Dimensions} from 'react-native'
import PropTypes from 'prop-types'
import ProgressBar from "react-native-form-progress"
import {
  DateTyper, 
  CheckBoxes, 
  PickerAtPosition, 
  SingleImageSelector,
  MultipleImageSelector, 
  Otp,
  InputText, 
  AutoComplete
} from './formComponent/index'
import Spinner from 'react-native-loading-spinner-overlay'

 const SignupFormComponent = (props) => {

  const {
    keyboardAutoOpenForText, 
    inputFields, 
    buttonSelectedStyle, 
    progressBar,  
    backgroundViewColor, 
    defaultColor, 
    helperTextStyle, 
    globalButtonText,
    buttonSelectedTextStyle,
    textStyle,
    onButtonClick,
    errorStyle,
    defaultErrorMessage,
    buttonNotSelectedStyle,
    buttonNotSelectedTextStyle,
    backIconStyle,
    onFinish
  } = props

  // All the component 
  const [index, setIndex] = useState(0)
  const [payload, setPayloadData] = useState({})
  const [Loading, toggleLoadingData] = useState(false)
  const [Error, setErrorData] = useState({status: false, message: ''})
  const [buttonDisable, toggleButtonDisable] = useState(true)
  // Current Component based on indux
  const currentComponent = inputFields[index]
  const {
    key, 
    type, 
    label, 
    helper, 
    buttonText, 
    validator
  } = currentComponent
  let {required} = currentComponent
  const templateOptions =  currentComponent.templateOptions || {}
  // if no template options, initlalize an empty object
  const {
    templateStyle, 
    options, 
    multipleSelect, 
    // Image Props
    cropHeight,
    cropWidth,
    //---- Date Format
    dateFormat,
    //---- OTP 
    noOfTextInput, 
    componentProps,
    //---- AutoComplete 
    asyncFunction,
    loaderRequired,
    listViewStyle,
    listTextStyle
  }  = templateOptions

  // Button Touch 
  const usedButtonText =  buttonText || globalButtonText
  // Setting button touch style based on state/validation
  const userButtonStyle = required ? buttonNotSelectedStyle : buttonSelectedStyle
  const userButtonTextStyle = required ? buttonNotSelectedTextStyle: buttonSelectedTextStyle
  // Initially setting up button state based on input from the json (required)

  // Setting up/mutating props 

  // --- Validators --- 
  // Validator hook effect in the bottom 
  const inputValidatorSelector = () => {
    if (validator  === undefined && !payload[key] && required) return () => true
    if (validator) return validator
    return () => false
  }
  const inputValidator = inputValidatorSelector()

  // --- Progress bar ---
  progressBar.currentProgress = index 
  progressBar.totalNumberOfProgressBars = inputFields.length
  progressBar.colorOfProgressBar = progressBar.colorOfProgressBar || defaultColor

  // All Functions declareation 
  // --------------------------
  const onChangeHandler =  (data, errorMessage=null) => {
    // Error Message is usually passed from the child components (look ad date componenet example )
    if (!errorMessage && (data === '' || data)) { // Doing data = "" in case someone want to remove last char of a string
      const currentData = {...payload}
      currentData[key] = data
      setPayloadData(currentData)
    } else if (errorMessage) {
      setErrorData({status: true, message: errorMessage})
    }
    return
  }

  const getValueFromState  = async () => {
    if (index === inputFields.length - 1) {
      if (onFinish) return onFinish
      else return 
    }
    setErrorData({status: false, message: ''})
    toggleLoadingData(true)
    const currentValue = payload[key]
    try {
      const eventTrack = await onButtonClick(index, key, currentValue, payload)
      if (index < inputFields.length) setIndex(index + 1)
      toggleButtonDisable(true)
      return toggleLoadingData(false)
    } catch (error) {
      if (error.message) setErrorData({status: true, message: error.message})
      else setErrorData({status: false, message: defaultErrorMessage})
      return toggleLoadingData(false)
    }
  }

  const decreaseStateIndex = () => {
    if (index > 0) return setIndex(index - 1)
    return 
  }

  // React Hooks
  // --------------------
  // Fired whenever payload changes, for validation
  useEffect(() => {
    const currentData = payload[key]
    const checkValidation = inputValidator(index, key, currentData, payload) 
    toggleButtonDisable(checkValidation)
  },[payload, index])
  
  const mapSignUpComponents = {
    text: (
      <InputText  
        {...componentProps}
        textInputStyle={templateStyle}
        defaultColor={defaultColor}
        upsideEmit={text => onChangeHandler(text)}
        value={payload[key] ? `${payload[key]}` : ''} 
      />
    ),
    dateTyper: (
      <DateTyper
      textInputStyle={[{color: defaultColor, width: (Dimensions.get('window').width * 0.6)/8 }, styles.nextInputStyle, templateStyle]} 
      upsideEmit={onChangeHandler}
      dateFormat={dateFormat}
      value={payload[key] ? `${payload[key]}` : ''}/>
    ),
    checkboxes: options === undefined ? null : (
      <CheckBoxes 
      options={options}
      upsideEmit={onChangeHandler}
      defaultColor={defaultColor}
      />
    ),  
    picker: options === undefined ? null : (
      <PickerAtPosition 
      options={options}
      upsideEmit={onChangeHandler} 
      value={payload[key] ? `${payload[key]}` : ''}
      />
    ),
    image:  (
      <SingleImageSelector
      upsideEmit={onChangeHandler}
      value={{...payload[key]}}
      defaultColor={defaultColor}
      cropWidth={cropWidth}
      cropHeight={cropHeight}
      />
    ),
    images:  (
      <MultipleImageSelector
        upsideEmit={onChangeHandler}
        defaultColor={defaultColor}
        cropWidth={cropWidth}
        cropHeight={cropHeight}
        value={Array.isArray(payload[key]) ? [...payload[key]] : null}
      />
    ), 
    otp: noOfTextInput === undefined ? null : (
      <Otp
        noOfTextInput={noOfTextInput}
        upsideEmit={onChangeHandler}
        value={Array.isArray(payload[key]) ? [...payload[key]] : null}
        textInputStyle={[{color: defaultColor, width: (Dimensions.get('window').width * 0.6)/8 }, styles.nextInputStyle, templateStyle]} 
      />
    ),
    autoComplete: asyncFunction === undefined ? null : (
      <AutoComplete
        {...componentProps}
        upsideEmit={onChangeHandler}
        value={{...payload[key]}}
        defaultColor={defaultColor}
        asyncFunction={asyncFunction}
        loaderRequired={loaderRequired}
        listViewStyle={listViewStyle}
        listTextStyle={listTextStyle}
      />
    )
  }

  const renderComponent = mapSignUpComponents[type]

  return (
    <View style={{ flex: 1}}>
    <View style={{backgroundColor: backgroundViewColor, height: '100%'}}>  
    <Spinner
      visible={Loading} />
      <ProgressBar {...progressBar} />
      {/* Back button */}
        {index !== 0 ? (
          <TouchableOpacity onPress={decreaseStateIndex} style={styles.backButton}>
            <Text style={[styles.backButtonText, backIconStyle]}> {'<'} </Text> 
          </TouchableOpacity> ) : <TouchableOpacity style={styles.nonBackButton}></TouchableOpacity>}
        <View style={styles.renderComponentView}>
          <Text style={[{color: defaultColor}, styles.textStyling, textStyle]}> {label}</Text>
          {renderComponent}
           { Error.status ?  (<Text style={[styles.subText, styles.errorColor,  helperTextStyle]}>{Error.message}</Text>): null }
          <Text style={[styles.subText, errorStyle]}>{helper}</Text>
          <TouchableOpacity 
            style={[required && buttonDisable ? styles.defaultButtonNotSelected : {borderColor: defaultColor, color: defaultColor}, styles.defaultButton, userButtonStyle]}
            onPress={getValueFromState}
            disabled={required && buttonDisable}
            > 
            <Text style={[ required && buttonDisable? styles.defaultButtonNotSelected : {borderColor: defaultColor, color: defaultColor}, styles.defaultButtonText, userButtonTextStyle]}> {usedButtonText} </Text> 
          </TouchableOpacity>
      </View>
    </View>
    </View>
  )
}


const styles = StyleSheet.create({
  mainView: {
    flex: 1
  }, 
  btnDefualtStyle: {
    color: 'green',
    backgroundColor: 'white', 
    width: 200,
    height: 50
  }, 
  renderComponentView: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center', 
    justifyContent: 'center',
    marginTop: 30,
  }, 
  textStyling: {
    fontSize: 28,
    fontWeight: 'bold'
  }, 
  defaultTextInputStyle: {
    width: 'auto',
    marginTop: 50,
    padding: 5,
    justifyContent: 'center',
    borderBottomWidth: 1, 
    width: Dimensions.get('window').width * 0.8,
    fontSize: 24
  }, 
  subText: {
    fontWeight: "300", 
    marginTop: 50, 
    fontSize: 16,
    width: Dimensions.get('window').width * 0.8,
    textAlign: 'center',
    color: '#757575'
  }, 
  defaultButton: {
    marginTop: 50,
    borderRadius: 20,
    width: Dimensions.get('window').width * 0.8,
    borderWidth: 1, 
    padding: 5, 
    display: 'flex',
    flexDirection: 'row',
    justifyContent: "center",
  }, 
  defaultButtonText: {
    fontSize: 25,
    fontWeight: 'bold'
  }, 
  errorColor: {
    marginTop: 20, 
    color: 'red',
    fontSize: 14,
    fontWeight: 'bold'
  }, 
  nextInputStyle: {
    textAlign: 'center', 
    marginLeft: 5, 
    marginRight: 5, 
    fontSize: 20,
    borderBottomWidth: 1,
    marginTop: 50
  }, 
  defaultButtonNotSelected: {
    color: '#9E9E9E',
    borderColor: '#9E9E9E'
  }, 
  backButton: {
    display: 'flex', 
    flexDirection: 'row', 
    justifyContent: 'flex-start',
    marginTop: 10, 
    marginLeft: 10
  },
  backButtonText: {
    fontSize: 28,
    color: '#BDBDBD'
  },
  nonBackButton: {
    marginTop: 38 // Margin top here is 38 because our back button style have margin top 10 and font size 28 (10+28) 
  }
})



SignupFormComponent.propTypes = { 
  textStyle: PropTypes.object,
  inputFields: PropTypes.array.isRequired,
  progressBar: PropTypes.object, 
  helperTextStyle: PropTypes.object,
  backgroundViewColor:  PropTypes.string, 
  defaultColor: PropTypes.string,
  onFinish: PropTypes.func,
  //Buttons
  buttonSelectedStyle: PropTypes.object,
  globalButtonText: PropTypes.string,
  buttonSelectedTextStyle: PropTypes.object,
  onButtonClick: PropTypes.func.isRequired,
  buttonNotSelectedStyle: PropTypes.object,
  buttonNotSelectedTextStyle: PropTypes.object,
  //Errors
  errorStyle: PropTypes.object, 
  defaultErrorMessage: PropTypes.string,
  backIconStyle: PropTypes.object
}

SignupFormComponent.defaultProps = {
  keyboardAutoOpenForText: true,
  progressBar: {
    blink: false,
  },
  globalButtonText: 'Next',
  backgroundViewColor: 'white',
  defaultColor: 'black', 
  defaultErrorMessage: 'Sorry Something went wrong'
}



export default SignupFormComponent






